File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/calendar/CalendarFragment.java
package com.gnoemes.shikimoriapp.presentation.view.calendar;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.gnoemes.shikimoriapp.R;
import com.gnoemes.shikimoriapp.entity.calendar.presentation.CalendarItemViewModel;
import com.gnoemes.shikimoriapp.presentation.presenter.calendar.CalendarPresenter;
import com.gnoemes.shikimoriapp.presentation.view.calendar.adapter.CalendarAdapter;
import com.gnoemes.shikimoriapp.presentation.view.calendar.adapter.provider.CalendarAnimeResourceProvider;
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.BaseFragment;
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RouterProvider;
import com.gnoemes.shikimoriapp.presentation.view.common.widget.NetworkErrorView;
import com.gnoemes.shikimoriapp.utils.date.converter.DateTimeConverter;
import com.gnoemes.shikimoriapp.utils.imageloader.ImageLoader;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;

public class CalendarFragment extends BaseFragment<CalendarPresenter, CalendarView>
        implements CalendarView {

    @BindView(R.id.refresh_layout)
    SwipeRefreshLayout refreshLayout;

    @BindView(R.id.list)
    RecyclerView recyclerView;

    @BindView(R.id.view_network_error)
    NetworkErrorView networkErrorView;

    @InjectPresenter
    CalendarPresenter presenter;
    @Inject
    DateTimeConverter dateTimeConverter;
    @Inject
    CalendarAnimeResourceProvider resourceProvider;
    @Inject
    ImageLoader imageLoader;
    private CalendarAdapter calendarAdapter;

    @ProvidePresenter
    CalendarPresenter providePresenter() {
        presenter = presenterProvider.get();
        if (getParentFragment() != null) {
            presenter.setLocalRouter(((RouterProvider) getParentFragment()).getLocalRouter());
        }
        return presenter;
    }

    public static CalendarFragment newInstance() {
        Bundle args = new Bundle();
        CalendarFragment fragment = new CalendarFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initList();
    }

    private void initList() {
        calendarAdapter = new CalendarAdapter(dateTimeConverter,
                resourceProvider,
                imageLoader,
                id -> getPresenter().onAnimeClicked(id));

        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(calendarAdapter);

        refreshLayout.setColorSchemeColors(getResources().getColor(R.color.red));
        refreshLayout.setOnRefreshListener(() -> getPresenter().loadCalendarData());
    }

    ///////////////////////////////////////////////////////////////////////////
    // GETTERS
    ///////////////////////////////////////////////////////////////////////////

    /**
     * Returns presenter
     */
    @Override
    protected CalendarPresenter getPresenter() {
        return presenter;
    }

    @Override
    protected int getFragmentLayout() {
        return R.layout.fragment_calendar;
    }

    ///////////////////////////////////////////////////////////////////////////
    // MVP
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public void showCalendarData(List<CalendarItemViewModel> items) {
        recyclerView.setVisibility(View.VISIBLE);
        calendarAdapter.addNewItems(items);
    }

    @Override
    public void hideList() {
        recyclerView.setVisibility(View.GONE);
    }

    @Override
    public void showNetworkErrorView() {
        networkErrorView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideNetworkErrorView() {
        networkErrorView.setVisibility(View.GONE);
    }

    @Override
    public void onShowLoading() {
        refreshLayout.setRefreshing(true);
    }

    @Override
    public void onHideLoading() {
        refreshLayout.setRefreshing(false);
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/calendar/CalendarFragment.kt
package com.gnoemes.shikimoriapp.presentation.view.calendar

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.gnoemes.shikimoriapp.R
import com.gnoemes.shikimoriapp.entity.calendar.presentation.CalendarItemViewModel
import com.gnoemes.shikimoriapp.presentation.presenter.calendar.CalendarPresenter
import com.gnoemes.shikimoriapp.presentation.view.calendar.adapter.CalendarAdapter
import com.gnoemes.shikimoriapp.presentation.view.calendar.adapter.provider.CalendarAnimeResourceProvider
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.BaseFragment
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RouterProvider
import com.gnoemes.shikimoriapp.utils.date.converter.DateTimeConverter
import com.gnoemes.shikimoriapp.utils.ifNotNull
import com.gnoemes.shikimoriapp.utils.imageloader.ImageLoader
import com.gnoemes.shikimoriapp.utils.kotlin.gone
import com.gnoemes.shikimoriapp.utils.kotlin.visible
import com.gnoemes.shikimoriapp.utils.themeDrawable
import kotlinx.android.synthetic.main.fragment_calendar.*
import kotlinx.android.synthetic.main.layout_default_list.*
import javax.inject.Inject

class CalendarFragment : BaseFragment<CalendarPresenter, CalendarView>(), CalendarView {

    @Inject
    lateinit var imageLoader: ImageLoader

    @Inject
    lateinit var dateTimeConverter: DateTimeConverter

    @Inject
    lateinit var resourceProvider: CalendarAnimeResourceProvider

    @InjectPresenter
    lateinit var calendarPresenter: CalendarPresenter

    @ProvidePresenter
    fun providePresenter(): CalendarPresenter {
        calendarPresenter = presenterProvider.get()

        parentFragment.ifNotNull {
            presenter.setLocalRouter((it as RouterProvider).localRouter)
        }

        return calendarPresenter
    }

    companion object {
        @JvmStatic
        fun newInstance() = CalendarFragment()
    }

    private val calendarAdapter by lazy { CalendarAdapter(dateTimeConverter, resourceProvider, imageLoader, CalendarAdapter.AnimeListener { presenter.onAnimeClicked(it) }) }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
    }

    @SuppressLint("ResourceType")
    private fun initViews() {
        with(list) {
            layoutManager = LinearLayoutManager(context)
            adapter = calendarAdapter
        }

        toolbar.ifNotNull { toolbar ->
            with(toolbar) {
                val icon = context?.themeDrawable(R.drawable.ic_tune, R.attr.colorText)

                inflateMenu(R.menu.menu_calendar)
                menu?.getItem(0)?.icon = icon
                setNavigationOnClickListener { presenter.onBackPressed() }
                setOnMenuItemClickListener {
                    when (it.itemId) {
                        R.id.item_my_ongoings -> presenter.onFilterClicked()
                        else -> Unit
                    }
                    false
                }
            }
        }

        view_network_error.gone()


        with(refresh_layout) {
            setColorSchemeColors(Color.RED)
            setOnRefreshListener { presenter.loadCalendar() }
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // GETTERS
    ///////////////////////////////////////////////////////////////////////////

    override fun getPresenter(): CalendarPresenter = calendarPresenter

    @LayoutRes
    override fun getFragmentLayout(): Int = R.layout.fragment_calendar


    ///////////////////////////////////////////////////////////////////////////
    // MVP
    ///////////////////////////////////////////////////////////////////////////

    override fun showCalendarData(items: MutableList<CalendarItemViewModel>?) {
        list.visible()
        calendarAdapter.bindItems(items)
    }

    override fun hideList() = list.gone()

    override fun showNetworkErrorView() = view_network_error.visible()

    override fun hideNetworkErrorView() = view_network_error.gone()

    override fun onShowLoading() {
        refresh_layout.isRefreshing = true
    }

    override fun onHideLoading() {
        refresh_layout.isRefreshing = false
    }


    ///////////////////////////////////////////////////////////////////////////
    // INNER CLASS
    ///////////////////////////////////////////////////////////////////////////

//    inner class MangaPagerAdapter(
//            private val calendarAdapter: CalendarAdapter,
////            private val chaptersAdapter: ChaptersAdapter
//    ) : ViewStatePagerAdapter() {
//
//    }

}